### ISSUE TITLE
官方微调代码 lr 不衰减？？

### ISSUE THREAD

**Issue by txy6666yr** [QwenLM/Qwen#571]
使用官方微调代码全参数微调，数据长度11w，使用`finetune_ds.sh`文件，微调参数如下：
```
export CUDA_DEVICE_MAX_CONNECTIONS=1
# export CUDA_LAUNCH_BLOCKING=1
DIR=`pwd`
# ``（反引号）用于在命令行中执行命令并返回结果
pip install -r requirements.txt

GPUS_PER_NODE=8
NNODES=1
NODE_RANK=0
MASTER_ADDR=localhost
MASTER_PORT=6001

MODEL="/mnt/pfs-guan-ssai/nlu/data/tianxy/Qwen/model/Qwen-14B" # Set the path if you do not want to load from huggingface directly
# ATTENTION: specify the path to your training data, which should be a json file consisting of a list of conversations.
# See the section for finetuning in README for more information.
DATA="/mnt/pfs-guan-ssai/nlu/data/tianxy/Qwen/finetune/data/train2.json"

DISTRIBUTED_ARGS="
    --nproc_per_node $GPUS_PER_NODE \
    --nnodes $NNODES \
    --node_rank $NODE_RANK \
    --master_addr $MASTER_ADDR \
    --master_port $MASTER_PORT
"

torchrun $DISTRIBUTED_ARGS finetune.py \
    --model_name_or_path $MODEL \
    --data_path $DATA \
    --bf16 True \
    --output_dir output_qwen1 \
    --num_train_epochs 2 \
    --per_device_train_batch_size 4 \
    --per_device_eval_batch_size 1 \
    --gradient_accumulation_steps 16 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 1000 \
    --save_total_limit 5 \
    --learning_rate 5e-5 \
    --weight_decay 0.1 \
    --adam_beta2 0.95 \
    --warmup_ratio 0.01 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --report_to "none" \
    --model_max_length 4096 \
    --gradient_checkpointing True \
    --lazy_preprocess True \
    --deepspeed ds_config_zero3.json
```
为什么lr warmup以后不衰减呢？
```
{'loss': 1.6809, 'learning_rate': 0.0, 'epoch': 0.0}
{'loss': 1.6028, 'learning_rate': 2.1533827903669654e-05, 'epoch': 0.01}
{'loss': 1.621, 'learning_rate': 3.413030972429927e-05, 'epoch': 0.01}
{'loss': 1.4919, 'learning_rate': 4.306765580733931e-05, 'epoch': 0.02}
{'loss': 1.4389, 'learning_rate': 5e-05, 'epoch': 0.02}
{'loss': 1.743, 'learning_rate': 5e-05, 'epoch': 0.03}
{'loss': 1.5303, 'learning_rate': 5e-05, 'epoch': 0.03}
{'loss': 1.4579, 'learning_rate': 5e-05, 'epoch': 0.04}
^A{'loss': 1.4574, 'learning_rate': 5e-05, 'epoch': 0.04}
{'loss': 1.4413, 'learning_rate': 5e-05, 'epoch': 0.04}
{'loss': 1.4693, 'learning_rate': 5e-05, 'epoch': 0.05}
{'loss': 1.364, 'learning_rate': 5e-05, 'epoch': 0.05}
{'loss': 1.2787, 'learning_rate': 5e-05, 'epoch': 0.06}
{'loss': 1.3077, 'learning_rate': 5e-05, 'epoch': 0.06}
{'loss': 1.2678, 'learning_rate': 5e-05, 'epoch': 0.07}
{'loss': 1.2351, 'learning_rate': 5e-05, 'epoch': 0.07}
{'loss': 1.2137, 'learning_rate': 5e-05, 'epoch': 0.08}
{'loss': 1.1881, 'learning_rate': 5e-05, 'epoch': 0.08}
{'loss': 1.1626, 'learning_rate': 5e-05, 'epoch': 0.08}
{'loss': 1.1735, 'learning_rate': 5e-05, 'epoch': 0.09}
{'loss': 1.1837, 'learning_rate': 5e-05, 'epoch': 0.09}
{'loss': 1.1669, 'learning_rate': 5e-05, 'epoch': 0.1}
{'loss': 1.1757, 'learning_rate': 5e-05, 'epoch': 0.1}
{'loss': 1.1658, 'learning_rate': 5e-05, 'epoch': 0.11}
{'loss': 1.1519, 'learning_rate': 5e-05, 'epoch': 0.11}
{'loss': 1.1623, 'learning_rate': 5e-05, 'epoch': 0.12}
{'loss': 1.1697, 'learning_rate': 5e-05, 'epoch': 0.12}
{'loss': 1.1643, 'learning_rate': 5e-05, 'epoch': 0.12}
{'loss': 1.1467, 'learning_rate': 5e-05, 'epoch': 0.13}
{'loss': 1.1431, 'learning_rate': 5e-05, 'epoch': 0.13}
{'loss': 1.1706, 'learning_rate': 5e-05, 'epoch': 0.14}
{'loss': 1.1373, 'learning_rate': 5e-05, 'epoch': 0.14}
{'loss': 1.14, 'learning_rate': 5e-05, 'epoch': 0.15}
{'loss': 1.1709, 'learning_rate': 5e-05, 'epoch': 0.15}
{'loss': 1.158, 'learning_rate': 5e-05, 'epoch': 0.15}
{'loss': 1.1811, 'learning_rate': 5e-05, 'epoch': 0.16}
{'loss': 1.1552, 'learning_rate': 5e-05, 'epoch': 0.16}
{'loss': 1.2, 'learning_rate': 5e-05, 'epoch': 0.17}
{'loss': 1.131, 'learning_rate': 5e-05, 'epoch': 0.17}
{'loss': 1.1405, 'learning_rate': 5e-05, 'epoch': 0.18}
{'loss': 1.1524, 'learning_rate': 5e-05, 'epoch': 0.18}
{'loss': 1.1513, 'learning_rate': 5e-05, 'epoch': 0.19}
{'loss': 1.1529, 'learning_rate': 5e-05, 'epoch': 0.19}
{'loss': 1.1066, 'learning_rate': 5e-05, 'epoch': 0.19}
{'loss': 1.1505, 'learning_rate': 5e-05, 'epoch': 0.2}
{'loss': 1.1216, 'learning_rate': 5e-05, 'epoch': 0.2}
{'loss': 1.1324, 'learning_rate': 5e-05, 'epoch': 0.21}
{'loss': 1.1558, 'learning_rate': 5e-05, 'epoch': 0.21}
{'loss': 1.1528, 'learning_rate': 5e-05, 'epoch': 0.22}
{'loss': 1.1389, 'learning_rate': 5e-05, 'epoch': 0.22}
{'loss': 1.1052, 'learning_rate': 5e-05, 'epoch': 0.23}
```

**Comment by supdizh** [comment]
+1，
好像是因为全参用的deepspeed，lrschedule用的WarmupLR吧
deepspeed有cosine lr scheduler吗？

**Comment by txy6666yr** [comment]
> +1， 好像是因为全参用的deepspeed，lrschedule用的WarmupLR吧 deepspeed有cosine lr scheduler吗？ OneCycle ? https://discuss.huggingface.co/t/learning-rate-with-deepspeed-is-fixed-despite-lr-set-to-auto/52053/3
嗯嗯 loss是正常掉的，我就没管

**Comment by BowieHsu** [comment]
在zero3.json里面把scheduler替换成如下格式:
"scheduler": {
	    "type": "WarmupDecayLR",
	    "params": {
		"warmup_min_lr": 0,
		"warmup_max_lr": "auto",
		"warmup_num_steps": "auto",
		"warmup_type": "linear",
		"total_num_steps": "auto"
	      }
	}
会在warmup之后开始以线性速度下降learning_rate，然后cosine的下降方式好像deepspeed没有支持，我目前先换了一个这个配置来进行训练，自己数据上测试针对chat模型以1e-5的learning_rate进行sft之后的效果比较差，learning_rate下降版本训好之后我对比一下 @txy6666yr

**Comment by txy6666yr** [comment]
> 在zero3.json里面把scheduler替换成如下格式: "scheduler": { "type": "WarmupDecayLR", "params": { "warmup_min_lr": 0, "warmup_max_lr": "auto", "warmup_num_steps": "auto", "warmup_type": "linear", "total_num_steps": "auto" } } 会在warmup之后开始以线性速度下降learning_rate，然后cosine的下降方式好像deepspeed没有支持，我目前先换了一个这个配置来进行训练，自己数据上测试针对chat模型以1e-5的learning_rate进行sft之后的效果比较差，learning_rate下降版本训好之后我对比一下 @txy6666yr

对 我看其他的项目deepspeed zero3 也是线性的，谢谢啦

**Comment by lucasjinreal** [comment]
到底哪一个能复现官方的？

**Comment by jklj077** [comment]
请参考这里的说明 https://github.com/QwenLM/Qwen/issues/672#issuecomment-1833405442

这一问题与DeepSpeed不支持cosine，transformers不完全支持混用deepspeed optimizer与transformers scheduler有关
